tools/third_party.py

```diff
@@ -0,0 +1,164 @@
+rust_crates_path = tp("rust_crates")
+
+
+# This function creates or modifies an environment so that it matches the
+# expectations of various google tools (gn, gclient, etc).
+def google_env(env=None, merge_env={}, depot_tools_path=depot_tools_path):
+    env = make_env(env=env, merge_env=merge_env)
+    # Depot_tools to be in the PATH, before Python.
+    path_prefix = depot_tools_path + os.path.pathsep
+    if not env['PATH'].startswith(path_prefix):
+        env['PATH'] = path_prefix + env['PATH']
+    # We're not using Google's internal infrastructure.
+    if os.name == 'nt' and not 'DEPOT_TOOLS_WIN_TOOLCHAIN' in env:
+        env['DEPOT_TOOLS_WIN_TOOLCHAIN'] = "0"
+    return env
+
+
+def fix_symlinks():
+    # Ensure the third_party directory exists.
+    try:
+        os.makedirs(third_party_path)
+    except:
+        pass
+
+    # Make symlinks to Yarn metadata living in the root repo.
+    remove_and_symlink("../package.json", tp("package.json"))
+    remove_and_symlink("../yarn.lock", tp("yarn.lock"))
+
+    # TODO(ry) Is it possible to remove these symlinks?
+    remove_and_symlink("v8/third_party/googletest", tp("googletest"), True)
+    remove_and_symlink("v8/third_party/jinja2", tp("jinja2"), True)
+    remove_and_symlink("v8/third_party/llvm-build", tp("llvm-build"), True)
+    remove_and_symlink("v8/third_party/markupsafe", tp("markupsafe"), True)
+
+    # On Windows, git doesn't create the right type of symlink if the symlink
+    # and it's target are in different repos. Here we fix the symlinks that exist
+    # in the root repo while their target is in the third_party repo.
+    remove_and_symlink("third_party/node_modules", root("node_modules"), True)
+    remove_and_symlink("third_party/v8/build", root("build"), True)
+    remove_and_symlink("third_party/v8/buildtools", root("buildtools"), True)
+    remove_and_symlink("third_party/v8/build_overrides",
+                       root("build_overrides"), True)
+    remove_and_symlink("third_party/v8/testing", root("testing"), True)
+
+
+# Run Yarn to install JavaScript dependencies.
+def run_yarn():
+    run(["yarn"], cwd=third_party_path)
+
+
+# Run Cargo to install Rust dependencies.
+def run_cargo():
+    # Deletes the cargo index lockfile; it appears that cargo itself doesn't do it.
+    # If the lockfile ends up in the git repo, it'll make cargo hang for everyone
+    # else who tries to run sync_third_party.
+    def delete_lockfile():
+        lockfiles = find_exts(
+            path.join(rust_crates_path, "registry/index"), '.cargo-index-lock')
+        for lockfile in lockfiles:
+            os.remove(lockfile)
+
+    # Delete the index lockfile in case someone accidentally checked it in.
+    delete_lockfile()
+
+    run(["cargo", "fetch", "--manifest-path=" + root("Cargo.toml")],
+        cwd=third_party_path,
+        merge_env={'CARGO_HOME': rust_crates_path})
+
+    # Delete the lockfile again so it doesn't end up in the git repo.
+    delete_lockfile()
```

Comment:
Is it possible to just tell cargo to not create a lockfile?

Side comment: Has the change to installing with cargo vs gclient been positive? It seems to do a lot of random unnecessary stuff.

Reply:
I didn't look for that

Reply:
> Side comment: Has the change to installing with cargo vs gclient been positive? It seems to do a lot of random unnecessary stuff.

I haven't added any new rust creates myself, so I basically don't have the right experience.

Reply:
--frozen            Require Cargo.lock and cache are up to date
        --locked            Require Cargo.lock is up to date
^-- maybe one of these two?

Reply:
I tried, I tried these flags, the same happens:

```
cargo fetch --frozen --manifest-path=C:\Users\BertBelder\d\deno\Cargo.toml
    Blocking waiting for file lock on the registry index
```

Reply:
What if we just .gitignore it?

Reply:
That doesn't work when you use this stanza to commit to the deno_third_party repo.
https://github.com/piscisaureus/deno/blob/4d08bb85a4276292c2121b221b4840865a13cd42/tools/sync_third_party.py#L3-L5